src/parser/shared/modules/AlwaysBeCasting.tsx

```diff
@@ -147,10 +84,139 @@ class AlwaysBeCasting extends Analyzer {
+      );
+      return;
+    }
+    if (start === end) {
+      // When a spell is made instant from a proc, it often shows as a zero-time channel.
+      // We can safely ignore these when calculating active time, as we'll capture the actual
+      // active time through the GCD.
+      return;
+    }
+
+    this.activeTimeEdges.push({ timestamp: start, value: 1 });
+    this.activeTimeEdges.push({ timestamp: end, value: -1 });
+    /*
+     * segements and total active time need to be regenerated after data is added,
+     * but won't actually be computed until queried (hopefully at end of fight)
+     */
+    this.workingActiveTimeSegments = undefined;
+    this.activeTimeMemo = undefined;
+  }
+
+  /**
+   * Generates working active time segements if needed. Most likely no one queries the
+   * activeTimeSegments until after processing is complete, in which case this function will be
+   * called only once.
+   *
+   * Algorithm for turning edges into segments:
+   * activeTimeEdges represent when arbitrarily overlapping individual active time segments start
+   * or stop. If we sort them in chronological order and then step through while incrementing the
+   * counter an activity start edge and decrementing the counter on an activity end edge, we can
+   * detect inactivity time (when counter is 0) and activity time (when counter is greater than 0).
+   * Use this to generate the segment's union, which will be non-overlapping and in order.
+   */
+  private checkAndGenerateActiveTimeSegments(): ActivitySegment[] {
+    if (this.workingActiveTimeSegments !== undefined) {
+      return this.workingActiveTimeSegments;
+    }
+
+    this.activeTimeEdges.sort((a, b) => a.timestamp - b.timestamp);
+
+    let activityCount = 0;
+    let activityStartTimestamp = 0;
+    this.workingActiveTimeSegments = [];
+    for (const e of this.activeTimeEdges) {
+      if (activityCount === 0 && e.value === 1) {
+        // upwards edge - activity started
+        activityStartTimestamp = e.timestamp;
+      } else if (activityCount === 1 && e.value === -1) {
+        // downwards edge - activity ended
+        this.workingActiveTimeSegments.push({ start: activityStartTimestamp, end: e.timestamp });
+      }
+      activityCount += e.value;
+    }
+
+    return this.workingActiveTimeSegments;
   }
 
+  /////////////////////////////////////////////////////////////////////////////
+  // PUBLIC GETTERS
+  //
+
+  get activeTimeSegments() {
+    return this.checkAndGenerateActiveTimeSegments();
+  }
+
+  /** The active time (in ms) recorded */
+  get activeTime() {
+    if (this.activeTimeMemo === undefined) {
+      this.activeTimeMemo = this.getActiveTimeMillisecondsInWindow(
+        this.owner.fight.start_time,
+        this.owner.fight.end_time,
+      );
+    }
+    return this.activeTimeMemo;
+  }
```

Comment:
does this break if you change the window being considered? i think probably not because the CLP gets rebuilt? i think?

Reply:
I think the play here is to also check the memo against start and end time, and if either of those change we clear the memo.